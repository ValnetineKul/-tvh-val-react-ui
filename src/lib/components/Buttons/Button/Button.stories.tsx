import React from 'react';
import { createMeta, createTemplate } from '../../../story-utils';
import { buttonBaseStoryArgs } from '../../ButtonBase';
import { AngleLeft, AngleRight } from '../../Icon/icons/functional';
import Icon from '../../Icon';
import Button from './Button';

const Template = createTemplate(Button);

const Primitive = Template.bind({});

Primitive.args = {
  label: 'Label',
  variant: 'primary',
  type: 'button',
  size: 'md',
  fullWidth: false,
  disabled: false,
  isLoading: false,
  buttonProps: { 'data-testid': 'test' },
};

export const None = Template.bind({});
None.args = {
  ...Primitive.args,
};

None.storyName = 'none';

export const StartIcon = Template.bind({});
StartIcon.args = {
  ...Primitive.args,
  startIcon: <Icon icon={AngleLeft} />,
};

StartIcon.storyName = 'startIcon';

export const EndIcon = Template.bind({});
EndIcon.args = {
  ...Primitive.args,
  endIcon: <Icon icon={AngleRight} />,
};

EndIcon.storyName = 'endIcon';

export default createMeta({
  component: Button,
  title: 'Components/Buttons/Button',
  argTypes: {
    onClick: {
      action: { argTypesRegex: '^on.*' },
      control: { disable: true },
      table: {
        type: {
          summary: '(event: any) => void',
        },
      },
    },
    startIcon: {
      control: {
        disable: true,
      },
      table: {
        type: { summary: 'Icon with FontAwesome definition' },
      },
    },
    endIcon: {
      control: {
        disable: true,
      },
      table: {
        type: { summary: 'Icon FontAwesome definition' },
      },
    },
    ...buttonBaseStoryArgs,
  },
});
