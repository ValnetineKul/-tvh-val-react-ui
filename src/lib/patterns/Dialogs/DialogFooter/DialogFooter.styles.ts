import { makeStyles } from '../../../../themes/core';

const useStyles = makeStyles({ name: 'DialogFooter' })((theme) => ({
  wrapper: {
    display: 'block',
  },
  list: {
    display: 'flex',
    listStyle: 'none',
    margin: theme.layoutSpacing['spacing/000'],
    padding: theme.layoutSpacing['spacing/000'],
  },
  directionHorizontal: {
    flexDirection: 'row-reverse',
    justifyContent: 'flex-start',
  },
  directionVertical: {
    flexDirection: 'column',
  },
  listItem: {
    margin: theme.layoutSpacing['spacing/300'],
  },
  alternativeActionHorizontal: {
    marginRight: 'auto',
    marginLeft: theme.layoutSpacing['spacing/300'],
    display: 'flex',
    alignItems: 'center',
  },
  alternativeActionVertical: {
    display: 'flex',
    margin: theme.layoutSpacing['spacing/300'],
    justifyContent: 'center',
  },
}));

export default useStyles;
