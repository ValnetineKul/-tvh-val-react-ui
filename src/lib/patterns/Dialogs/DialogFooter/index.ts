import DialogFooter from './DialogFooter';

export type { DialogFooterProps } from './DialogFooter';
export default DialogFooter;
