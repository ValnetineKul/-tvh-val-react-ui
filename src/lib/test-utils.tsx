import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
