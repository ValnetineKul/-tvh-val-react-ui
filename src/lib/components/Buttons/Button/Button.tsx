import { default as MuiButton } from '@mui/material/Button';
import type { ReactElement, MouseEventHandler, KeyboardEvent } from 'react';
import React, { forwardRef } from 'react';
import { capitalize } from '@mui/material/utils';
import type { ButtonBaseProps } from '../../ButtonBase';
import Surface from '../../Surfaces/Surface';
import type { Surface as SurfaceColor } from '../../../../themes/core';
import { useSurface } from '../../../../themes/core';
import SpinnerWithBackdrop from '../../ProgressIndicators/Spinner/SpinnerWithBackdrop';
import useStyles from './Button.styles';
import useGetButtonProps from './helpers/useGetButtonProps';
import type { DataAttributes } from '../../../types/common';

export interface ButtonCommonProps extends ButtonBaseProps {
  variant?: 'primary' | 'secondary' | 'tertiary' | 'link';
  type?: 'button' | 'submit';
  linkType?: 'default' | 'primary';
  label: string | ReactElement;
  size?: 'sm' | 'md' | 'lg';
  disabled?: boolean;
  isLoading?: boolean;
  fullWidth?: boolean;
  className?: string;
  surfaceClassName?: string;
  containerClassName?: string;
  onClick?: MouseEventHandler<HTMLButtonElement>;
  onKeyPress?: (event: KeyboardEvent<HTMLButtonElement>) => void;
  buttonProps?: DataAttributes;
}

interface NoIcon {
  startIcon?: never;
  endIcon?: never;
}
interface StartIcon {
  startIcon: ReactElement;
  endIcon?: never;
}
interface EndIcon {
  startIcon?: never;
  endIcon: ReactElement;
}

export type ButtonIconProps = NoIcon | StartIcon | EndIcon;

export type ButtonProps = ButtonCommonProps & ButtonIconProps;

const Button = forwardRef<HTMLButtonElement | HTMLAnchorElement, ButtonProps>(
  (
    {
      variant = 'primary',
      type = 'button',
      label,
      onClick,
      size = 'md',
      fullWidth = false,
      isLoading = false,
      className,
      surfaceClassName,
      containerClassName,
      ...props
    },
    ref
  ) => {
    const { startIcon, endIcon, linkType, buttonProps, ...restProps } = props;

    const { isLink, shouldShowLoader, buttonBaseProps } = useGetButtonProps(isLoading, restProps, variant);
    const shouldShowFullWidth = fullWidth && variant !== 'link';

    const { color: onSurface } = useSurface();

    let variantColor: SurfaceColor;
    switch (variant) {
      case 'primary':
        variantColor = 'Primary';
        break;
      case 'secondary':
        variantColor = 'Secondary';
        break;
      case 'tertiary':
        variantColor = onSurface;
        break;
      case 'link':
        variantColor = onSurface;
        break;
      default:
        variantColor = 'Primary';
    }

    const { classes, cx } = useStyles({ variantColor });
    const sizeClassName = variant !== 'link' && (classes as Record<string, string>)[`size${capitalize(size)}`];

    return (
      <Surface
        color={variantColor}
        className={cx(classes.surface, { [classes.fullWidth]: shouldShowFullWidth }, surfaceClassName)}
      >
        <SpinnerWithBackdrop
          className={cx(
            { [classes.tertiaryButtonBackdrop]: variant === 'tertiary', [classes.fullWidth]: shouldShowFullWidth },
            containerClassName
          )}
          isLoading={shouldShowLoader}
          isBackdropOpaque
        >
          <MuiButton
            type={isLink ? undefined : type}
            className={cx(
              classes.root,
              sizeClassName,
              {
                [classes.button]: variant !== 'link',
                [classes.tertiaryButton]: variant === 'tertiary',
                [classes.variantLink]: variant === 'link',
                [classes.linkTypeDefault]: variant === 'link' && linkType !== 'primary',
                [classes.linkTypePrimary]: variant === 'link' && linkType === 'primary',
              },
              className
            )}
            classes={{
              startIcon: classes.startIcon,
              endIcon: classes.endIcon,
            }}
            startIcon={startIcon}
            endIcon={endIcon}
            fullWidth={fullWidth}
            onClick={onClick}
            disableElevation
            ref={ref}
            {...restProps}
            {...buttonBaseProps}
            {...buttonProps}
          >
            {label}
          </MuiButton>
        </SpinnerWithBackdrop>
      </Surface>
    );
  }
);
Button.displayName = 'Button';

export default Button;
