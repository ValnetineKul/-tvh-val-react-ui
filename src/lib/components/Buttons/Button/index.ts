import Button from './Button';
import type { ButtonProps, ButtonCommonProps, ButtonIconProps } from './Button';

export default Button;
export type { ButtonProps, ButtonCommonProps, ButtonIconProps };
