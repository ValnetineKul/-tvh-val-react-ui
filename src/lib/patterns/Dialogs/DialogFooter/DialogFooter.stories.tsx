import { action } from '@storybook/addon-actions';
import React from 'react';
import Button from '../../../components/Buttons/Button';
import { createMeta, createTemplate } from '../../../story-utils';
import DialogFooter from './DialogFooter';

const Template = createTemplate(DialogFooter);

export const Primitive = Template.bind({});
Primitive.args = {
  actionButtons: [<Button variant="primary" label="Primary action" />, <Button variant="tertiary" label="Cancel" />],
  alternativeAction: 'AlternativeAction',
  onAlternativeActionClick: action('onAlternativeActionClick'),
};

Primitive.storyName = 'DialogFooter';

export default createMeta({
  component: DialogFooter,
  title: 'Patterns/Dialogs/DialogFooter',
  argTypes: {
    actionButtons: {
      control: {
        disable: true,
      },
      table: {
        type: { summary: 'The action buttons of the component' },
      },
    },
    component: {
      control: {
        disable: true,
      },
    },
    onAlternativeActionClick: { control: { disable: true } },
  },
});
