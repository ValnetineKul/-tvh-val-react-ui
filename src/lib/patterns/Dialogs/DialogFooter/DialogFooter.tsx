import DialogActions from '@mui/material/DialogActions';

import type { FC } from 'react';
import React from 'react';
import { capitalize } from '@mui/material/utils';
import useScreenSize from '../../../hooks/useScreenSize';
import Divider from '../../../components/Divider';
import Button from '../../../components/Buttons/Button';
import type { ButtonBaseProps } from '../../../components/ButtonBase';
import { useButtonBase } from '../../../components/ButtonBase';
import useStyles from './DialogFooter.styles';
import type { DataAttributes } from '../../../types/common';

type Direction = 'horizontal' | 'vertical';
export interface DialogFooterProps extends ButtonBaseProps {
  direction?: Direction;
  actionButtons: React.ReactElement[];
  alternativeAction?: string;
  onAlternativeActionClick?: () => void;
  buttonProps?: React.HTMLAttributes<HTMLButtonElement | HTMLAnchorElement> & DataAttributes;
}

const DialogFooter: FC<DialogFooterProps> = ({
  direction: directionProp,
  actionButtons,
  alternativeAction,
  onAlternativeActionClick,
  buttonProps,
  ...props
}) => {
  const { classes, cx } = useStyles();
  const buttonBaseProps = useButtonBase(props);
  const { isTabletUp } = useScreenSize('Tablet');

  const direction = ((): Direction => {
    if (directionProp) return directionProp;
    return isTabletUp ? 'horizontal' : 'vertical';
  })();

  const directionClassName = (classes as Record<string, string>)[`direction${capitalize(direction)}`];
  const alternativeActionClassName = (classes as Record<string, string>)[`alternativeAction${capitalize(direction)}`];

  return (
    <>
      <Divider direction="horizontal" />
      <DialogActions className={classes.wrapper}>
        <ul className={cx(classes.list, directionClassName)}>
          {actionButtons.map((item: React.ReactElement, key) => (
            <li className={classes.listItem} key={key}>
              {direction === 'horizontal' ? item : React.cloneElement(item, { fullWidth: true })}
            </li>
          ))}
          {alternativeAction && (
            <li className={alternativeActionClassName}>
              <Button
                label={alternativeAction}
                variant="link"
                onClick={onAlternativeActionClick}
                {...buttonBaseProps}
                {...buttonProps}
              />
            </li>
          )}
        </ul>
      </DialogActions>
    </>
  );
};

export default DialogFooter;
