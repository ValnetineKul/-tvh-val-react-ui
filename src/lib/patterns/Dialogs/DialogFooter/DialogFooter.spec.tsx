import React from 'react';
import { render, userEvent } from '../../../test-utils';
import DialogFooter from './DialogFooter';
import Button from '../../../components/Buttons/Button/Button';

const props = {
  actionButtons: [<Button variant="primary" label="Primary" />],
};

describe('DialogFooter', () => {
  it('Should render correctly', () => {
    const { container } = render(<DialogFooter {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('Should render alternative action', () => {
    it('button', () => {
      const { getByRole } = render(<DialogFooter {...props} alternativeAction="AlternativeAction" />);
      const button = getByRole('button', { name: 'AlternativeAction' });
      expect(button).toBeInTheDocument();
      expect(button.className).toMatch('Button-variantLink');
    });

    it('link', () => {
      const { getByRole } = render(<DialogFooter {...props} alternativeAction="AlternativeAction" href="/test" />);
      const link = getByRole('link', { name: 'AlternativeAction' });
      expect(link).toBeInTheDocument();
      expect(link.className).toMatch('Button-variantLink');
    });

    it('with callback', () => {
      const onAlternativeActionClick = jest.fn();
      const { getByRole } = render(
        <DialogFooter
          {...props}
          alternativeAction="AlternativeAction"
          onAlternativeActionClick={onAlternativeActionClick}
        />
      );
      const button = getByRole('button', { name: 'AlternativeAction' });
      userEvent.click(button);
      expect(onAlternativeActionClick).toHaveBeenCalledTimes(1);
    });
  });

  describe('Should set modal action buttons direction', () => {
    const cases = [
      [undefined, 'DialogFooter-directionVertical'],
      ['horizontal', 'DialogFooter-directionHorizontal'],
      ['vertical', 'DialogFooter-directionVertical'],
    ];
    test.each(cases)(
      'given %p, returns %p',
      (input: React.ComponentProps<typeof DialogFooter>['direction'], expected: string) => {
        const { getByRole } = render(<DialogFooter direction={input} {...props} />);
        const buttonsList = getByRole('list');
        expect(buttonsList.className).toMatch(expected);
      }
    );
  });
});
