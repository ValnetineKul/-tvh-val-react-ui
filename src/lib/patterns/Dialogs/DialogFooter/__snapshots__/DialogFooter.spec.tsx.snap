// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialogFooter Should render correctly 1`] = `
<hr
  class="MuiDivider-root MuiDivider-fullWidth tss-1pam8iy-Divider-root-Divider-horizontal mui-9mgopn-MuiDivider-root"
/>
`;
